File: recyclerun-team/mobile/modules/live-vision-pipeline/android/src/main/java/expo/modules/livevisionpipeline/LiveVisionObjectDetectorFrameProcessorPlugin.kt
package expo.modules.livevisionpipeline

import android.graphics.Rect
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.objects.DetectedObject
import com.google.mlkit.vision.objects.ObjectDetection
import com.google.mlkit.vision.objects.defaults.ObjectDetectorOptions
import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin
import kotlin.math.abs
import kotlin.math.hypot
import kotlin.math.max
import kotlin.math.min

private data class TrackHistory(
  var firstSeenMs: Long,
  var lastSeenMs: Long,
  var stableSinceMs: Long,
  var stableFrames: Int,
  var lastRect: Rect
)

class LiveVisionObjectDetectorFrameProcessorPlugin(
  options: Map<String, Any>? = null
) : FrameProcessorPlugin() {
  private val detector = ObjectDetection.getClient(
    ObjectDetectorOptions.Builder()
      .setDetectorMode(ObjectDetectorOptions.STREAM_MODE)
      .enableMultipleObjects()
      .enableClassification()
      .build()
  )
  private val histories = LinkedHashMap<String, TrackHistory>()
  private val stableFramesThreshold = ((options?.get("stableFramesThreshold") as? Number)?.toInt() ?: 2).coerceAtLeast(1)
  private val stableMsThreshold = ((options?.get("stableMsThreshold") as? Number)?.toLong() ?: 900L).coerceAtLeast(100L)
  private val motionThreshold = ((options?.get("motionThreshold") as? Number)?.toDouble() ?: 0.08).coerceAtLeast(0.0)

  override fun callback(frame: Frame, params: Map<String, Any>?): Any {
    val startedAtNs = System.nanoTime()
    val image = frame.image
    val imageProxy = frame.imageProxy
    val inputImage = InputImage.fromMediaImage(image, imageProxy.imageInfo.rotationDegrees)
    val frameTimestampMs = frame.timestamp / 1_000_000.0
    val width = frame.width.toDouble().coerceAtLeast(1.0)
    val height = frame.height.toDouble().coerceAtLeast(1.0)

    val detectedObjects = try {
      Tasks.await(detector.process(inputImage))
    } catch (_: Throwable) {
      emptyList<DetectedObject>()
    }

    val activeKeys = HashSet<String>()
    val detections = detectedObjects.take(5).map { detectedObject ->
      val trackKey = buildTrackKey(detectedObject)
      activeKeys.add(trackKey)
      val rect = detectedObject.boundingBox
      val areaScore = ((rect.width().toDouble() / width) * (rect.height().toDouble() / height)).coerceIn(0.0, 1.0)
      val history = histories[trackKey]
      val motionScore = history?.let { calculateMotionScore(it.lastRect, rect, width, height) } ?: 0.0
      val nowMs = frameTimestampMs.toLong()
      val nextHistory = if (history == null) {
        TrackHistory(
          firstSeenMs = nowMs,
          lastSeenMs = nowMs,
          stableSinceMs = nowMs,
          stableFrames = 1,
          lastRect = Rect(rect)
        )
      } else {
        if (motionScore <= motionThreshold) {
          history.stableFrames += 1
        } else {
          history.stableFrames = 1
          history.stableSinceMs = nowMs
        }
        history.lastSeenMs = nowMs
        history.lastRect = Rect(rect)
        history
      }
      histories[trackKey] = nextHistory

      val stableMs = max(0L, nowMs - nextHistory.stableSinceMs)
      val primaryLabel = detectedObject.labels.firstOrNull()
      mapOf(
        "track_id" to trackKey,
        "bbox_norm" to mapOf(
          "x" to (rect.left.toDouble() / width).coerceIn(0.0, 1.0),
          "y" to (rect.top.toDouble() / height).coerceIn(0.0, 1.0),
          "width" to (rect.width().toDouble() / width).coerceIn(0.0, 1.0),
          "height" to (rect.height().toDouble() / height).coerceIn(0.0, 1.0),
        ),
        "first_seen_ms" to nextHistory.firstSeenMs.toDouble(),
        "last_seen_ms" to nextHistory.lastSeenMs.toDouble(),
        "stable_frames" to nextHistory.stableFrames.toDouble(),
        "stable_ms" to stableMs.toDouble(),
        "motion_score" to motionScore.coerceIn(0.0, 1.0),
        "area_score" to areaScore,
        "local_confidence" to ((primaryLabel?.confidence ?: 0.0f).toDouble()).coerceIn(0.0, 1.0),
        "coarse_label" to ((primaryLabel?.text ?: "").lowercase()),
        "is_stable" to (nextHistory.stableFrames >= stableFramesThreshold && stableMs >= stableMsThreshold)
      )
    }

    pruneInactiveTracks(activeKeys, frameTimestampMs.toLong())

    val detectorLatencyMs = (System.nanoTime() - startedAtNs) / 1_000_000.0
    return mapOf(
      "timestamp_ms" to frameTimestampMs,
      "width" to width,
      "height" to height,
      "detector_latency_ms" to detectorLatencyMs,
      "detections" to detections
    )
  }

  private fun buildTrackKey(detectedObject: DetectedObject): String {
    val trackingId = detectedObject.trackingId
    if (trackingId != null) {
      return "track_$trackingId"
    }
    val rect = detectedObject.boundingBox
    val label = detectedObject.labels.firstOrNull()?.text?.lowercase() ?: "object"
    return "track_${label}_${rect.left}_${rect.top}_${rect.width()}_${rect.height()}"
  }

  private fun calculateMotionScore(previous: Rect, current: Rect, width: Double, height: Double): Double {
    val prevCx = previous.exactCenterX().toDouble() / width
    val prevCy = previous.exactCenterY().toDouble() / height
    val currCx = current.exactCenterX().toDouble() / width
    val currCy = current.exactCenterY().toDouble() / height
    val centerDelta = hypot(currCx - prevCx, currCy - prevCy)
    val widthDelta = abs(current.width().toDouble() - previous.width().toDouble()) / width
    val heightDelta = abs(current.height().toDouble() - previous.height().toDouble()) / height
    return min(1.0, centerDelta + widthDelta + heightDelta)
  }

  private fun pruneInactiveTracks(activeKeys: Set<String>, nowMs: Long) {
    val iterator = histories.entries.iterator()
    while (iterator.hasNext()) {
      val entry = iterator.next()
      if (!activeKeys.contains(entry.key) && nowMs - entry.value.lastSeenMs > 2_500L) {
        iterator.remove()
      }
    }
  }
}

File: recyclerun-team/mobile/modules/live-vision-pipeline/ios/LiveVisionObjectDetectorFrameProcessorPlugin.swift
import Foundation
import MLKitObjectDetection
import MLKitVision
import VisionCamera

private final class LiveVisionTrackState {
  var firstSeenMs: Double
  var lastSeenMs: Double
  var stableSinceMs: Double
  var stableFrames: Int
  var lastRect: CGRect

  init(firstSeenMs: Double, lastSeenMs: Double, stableSinceMs: Double, stableFrames: Int, lastRect: CGRect) {
    self.firstSeenMs = firstSeenMs
    self.lastSeenMs = lastSeenMs
    self.stableSinceMs = stableSinceMs
    self.stableFrames = stableFrames
    self.lastRect = lastRect
  }
}

@objc(LiveVisionObjectDetectorFrameProcessorPlugin)
public final class LiveVisionObjectDetectorFrameProcessorPlugin: FrameProcessorPlugin {
  private let detector: ObjectDetector
  private var histories: [String: LiveVisionTrackState] = [:]
  private let stableFramesThreshold: Int
  private let stableMsThreshold: Double
  private let motionThreshold: Double

  public override init(proxy: VisionCameraProxyHolder, options: [AnyHashable: Any]? = nil) {
    let config = options ?? [:]
    stableFramesThreshold = max(1, (config["stableFramesThreshold"] as? NSNumber)?.intValue ?? 2)
    stableMsThreshold = max(100.0, (config["stableMsThreshold"] as? NSNumber)?.doubleValue ?? 900.0)
    motionThreshold = max(0.0, (config["motionThreshold"] as? NSNumber)?.doubleValue ?? 0.08)

    let detectorOptions = ObjectDetectorOptions()
    detectorOptions.detectorMode = .stream
    detectorOptions.shouldEnableMultipleObjects = true
    detectorOptions.shouldEnableClassification = true
    detector = ObjectDetector.objectDetector(options: detectorOptions)
    super.init(proxy: proxy, options: options)
  }

  public override func callback(_ frame: Frame!, withArguments arguments: [AnyHashable: Any]! = nil) -> Any! {
    let startedAt = CFAbsoluteTimeGetCurrent()
    let visionImage = VisionImage(buffer: frame.buffer)
    visionImage.orientation = frame.orientation
    let timestampMs = frame.timestamp / 1000.0
    let width = max(Double(frame.width), 1.0)
    let height = max(Double(frame.height), 1.0)

    let objects: [Object]
    do {
      objects = try detector.results(in: visionImage)
    } catch {
      objects = []
    }

    var activeTrackKeys = Set<String>()
    let detections: [[String: Any]] = objects.prefix(5).map { object in
      let trackKey = buildTrackKey(object)
      activeTrackKeys.insert(trackKey)
      let rect = object.frame
      let areaScore = max(0.0, min(1.0, (rect.width / CGFloat(width)) * (rect.height / CGFloat(height))))
      let history = histories[trackKey]
      let motionScore = history.map { calculateMotionScore(previous: $0.lastRect, current: rect, width: width, height: height) } ?? 0.0

      let nextHistory: LiveVisionTrackState
      if let history {
        if motionScore <= motionThreshold {
          history.stableFrames += 1
        } else {
          history.stableFrames = 1
          history.stableSinceMs = timestampMs
        }
        history.lastSeenMs = timestampMs
        history.lastRect = rect
        nextHistory = history
      } else {
        nextHistory = LiveVisionTrackState(
          firstSeenMs: timestampMs,
          lastSeenMs: timestampMs,
          stableSinceMs: timestampMs,
          stableFrames: 1,
          lastRect: rect
        )
      }
      histories[trackKey] = nextHistory

      let stableMs = max(0.0, timestampMs - nextHistory.stableSinceMs)
      let primaryLabel = object.labels.first
      return [
        "track_id": trackKey,
        "bbox_norm": [
          "x": max(0.0, min(1.0, rect.minX / CGFloat(width))),
          "y": max(0.0, min(1.0, rect.minY / CGFloat(height))),
          "width": max(0.0, min(1.0, rect.width / CGFloat(width))),
          "height": max(0.0, min(1.0, rect.height / CGFloat(height))),
        ],
        "first_seen_ms": nextHistory.firstSeenMs,
        "last_seen_ms": nextHistory.lastSeenMs,
        "stable_frames": nextHistory.stableFrames,
        "stable_ms": stableMs,
        "motion_score": max(0.0, min(1.0, motionScore)),
        "area_score": areaScore,
        "local_confidence": max(0.0, min(1.0, primaryLabel?.confidence ?? 0.0)),
        "coarse_label": primaryLabel?.text.lowercased() ?? "",
        "is_stable": nextHistory.stableFrames >= stableFramesThreshold && stableMs >= stableMsThreshold,
      ]
    }

    pruneInactiveTracks(activeTrackKeys: activeTrackKeys, nowMs: timestampMs)

    return [
      "timestamp_ms": timestampMs,
      "width": width,
      "height": height,
      "detector_latency_ms": (CFAbsoluteTimeGetCurrent() - startedAt) * 1000.0,
      "detections": detections,
    ]
  }

  private func buildTrackKey(_ object: Object) -> String {
    if object.trackingID != 0 {
      return "track_\(object.trackingID)"
    }
    let frame = object.frame
    let label = object.labels.first?.text.lowercased() ?? "object"
    return "track_\(label)_\(Int(frame.minX))_\(Int(frame.minY))_\(Int(frame.width))_\(Int(frame.height))"
  }

  private func calculateMotionScore(previous: CGRect, current: CGRect, width: Double, height: Double) -> Double {
    let prevCx = Double(previous.midX) / width
    let prevCy = Double(previous.midY) / height
    let currCx = Double(current.midX) / width
    let currCy = Double(current.midY) / height
    let centerDelta = hypot(currCx - prevCx, currCy - prevCy)
    let widthDelta = abs(Double(current.width - previous.width)) / width
    let heightDelta = abs(Double(current.height - previous.height)) / height
    return min(1.0, centerDelta + widthDelta + heightDelta)
  }

  private func pruneInactiveTracks(activeTrackKeys: Set<String>, nowMs: Double) {
    histories = histories.filter { key, value in
      activeTrackKeys.contains(key) || (nowMs - value.lastSeenMs) <= 2500.0
    }
  }
}
